client/src/escrow_client/mod.rs

```diff
@@ -1,155 +1,152 @@
+use std::sync::Arc;
+
+use crate::common::model::EscrowRegistration;
+use cdk::nuts::Token;
+
 use super::*;
-use cashu_escrow_common::nostr::PubkeyMessage;
-use cdk::nuts::PublicKey;
 
-pub enum Trader {
-    Buyer(EscrowUser),
-    Seller(EscrowUser),
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum TradeMode {
+    Buyer,
+    Seller,
 }
 
-pub struct EscrowUser {
-    pub escrow_coordinator_npub: String,
-    pub escrow_pk_ts: (PublicKey, Timestamp),
-    pub contract: TradeContract,
-    pub wallet: EcashWallet,
-    pub nostr_client: NostrClient,
+pub struct InitEscrowClient {
+    ecash_wallet: ClientEcashWallet,
+    escrow_contract: TradeContract,
+    trade_mode: TradeMode,
 }
 
-impl Trader {
-    pub async fn init_trade(&self) -> anyhow::Result<()> {
-        match self {
-            Trader::Buyer(config) => {
-                self.buyer_pipeline(config).await?;
-                Ok(())
-            }
-            Trader::Seller(config) => {
-                self.seller_pipeline(config).await?;
-                Ok(())
-            }
+/// Initial Escrow Client state.
+impl InitEscrowClient {
+    pub fn new(
+        ecash_wallet: ClientEcashWallet,
+        escrow_contract: TradeContract,
+        trade_mode: TradeMode,
+    ) -> Self {
+        Self {
+            ecash_wallet,
+            escrow_contract,
+            trade_mode,
         }
     }
 
-    async fn buyer_pipeline(&self, config: &EscrowUser) -> anyhow::Result<()> {
-        let token = config.wallet.create_escrow_token(config).await?;
-        dbg!("Sending token to the seller: {}", token.as_str());
-
-        config
-            .nostr_client
-            .submit_trade_token_to_seller(&config.contract.npub_seller, &token)
+    /// The trade initialization is the same for both buyer and seller.
+    ///
+    /// After this the coordinator data is set, state trade registered.
+    ///
+    /// After this state the trade contract is effectfull as well, possible coordinator fees must be payed.
+    pub async fn register_trade(
+        &self,
+        nostr_client: Arc<NostrClient>,
+    ) -> anyhow::Result<RegisteredEscrowClient> {
+        let nostr_client_ref = nostr_client.clone();
+        let reg_msg_fut =
```

Comment:
I don't understand why we need to spawn a new task to listen to the escrow msg and then await it afterwards. Couldn't we just send the contract message to the coordinator and then start listening to answers (receive_escrow_message) of the coordinator with a limit higher than 0 (e.g. 1) so we receive "old" messages? 
```
        let message_filter = Filter::new()
            .kind(Kind::GiftWrap)
            .pubkey(self.keys.public_key())
            .limit(1);
```

Reply:
According to the maintainer of rust-nostr limit(0) is the way to go in case of GiftWrapped event and thus also for NIP-17. See the thread https://github.com/rust-nostr/nostr/issues/173#issuecomment-2292065627. Those events have a random created_at field which can be within 2 days in the past. If the client isn't  listening before sending the contract, in some situations it misses the answer of the coordinator. I have observed this several times.

Reply:
When you observed the client missing the events, did you test with Timestamp::now() or limit(0)? Just for my general understanding, I see why Timestamp::now() doesn't work due to the random created_at, but in my understanding limit(1) should still work as it just tells the relay to return the last stored event. In this case we would lose the "safety" of only receiving a event that was created after the contract started (current TS), but timestamps can be wrong anyways so we shouldn't rely on them.
Otherwise i would be wondering how clients like Amethyst handle NIP17 dms, as they are also not always listening. I also wonder if limits(1) returns the latest event by created_at (which would be quite random with NIP17) or by storage/receiving order.

Reply:
I've observed this issue for a long time, when we had the since(now) filter and also after switching to limit(0). It happened when the buyer sends first the contract and the seller as second one. In this case the coordinator fires almost always the registration before the buyer starts listening for events. You can see this for example on master and in earlier stages of this branch. The fix to this by listening before sending the contract to the coordinator happened way later than the switch to nip-17. Now we still have a similar issue in a later step in case of the seller, it can start listening for the token after the buyer sent the token, I have also observed this. I think the approach with a higher limit isn't reliable enough and we need a different design. The nostr client must start listening from the very beginning and buffer relevant messages, then the escrow client can ask any time for messages with a new version of receive_escrow_message and pick the one it is looking for. It is an idea similar to an actor system where the nostr client acts as a mailbox for the escrow client. Note that in this approach we wouldn't limit the filter, limit(0) would keep listening for new events.

Reply:
I think we could just split receiving in two parts. One function that subscribes to the filter and returns the subscription. And another function that pulls events from the subscription (.recv()) channel. So nostr-sdk is the buffer and we don't need extra tasks.

On the other side, Nostr clients like Amethyst also seem to be able to pull NIP17 events afterwards, i guess they just subscribe to NIP17 events, and then sort them by the created_at in the rumor.

I think we also should consider that traders are most probably not always online at the same time, and go offline between the single steps when waiting for the other party. So we may need to be able to continue at a specific step loading the state from somewhere.

Reply:
I like your idea of trying to use the nostr-sdk as event buffer, I'll try it out.

Reply:
I also think Amethyst does a kind of logic as you mentioned.

Reply:
Cool👍 
Yeah let's open an issue for this so we get this PR merged and can work on smaller issues in parallel.